File: C/src/dynamic_stack.c
#include <stdio.h>
#include <stdlib.h>
#include "dynamic_stack.h"

void inicializarDynamicStack(DynamicStack *stack) {
    stack->topo = NULL;
    stack->size = 0;
}

int empilharDynamicStack(DynamicStack *stack, Registro valor) {
    No *novo = (No*) malloc(sizeof(No));
    if (!novo) return 0;

    novo->valor = valor;
    novo->prox = stack->topo;
    stack->topo = novo;
    stack->size++;
    return 1;
}

int desempilharDynamicStack(DynamicStack *stack, Registro *valor) {
    if (pilhaVazia(stack)) return 0;

    No *remover = stack->topo;
    *valor = remover->valor;
    stack->topo = remover->prox;
    free(remover);
    stack->size--;
    return 1;
}

int pilhaVaziaDynamicStack(DynamicStack *stack) {
    return stack->size == 0;
}

void imprimirDynamicStack(DynamicStack *stack) {
    No *atual = stack->topo;
    while (atual != NULL) {
        printf("%d ", atual->valor.valor);
        atual = atual->prox;
    }
    printf("\n");
}

void salvarDynamicStackEmArquivo(DynamicStack *stack, const char *nomeArquivo) {
    FILE *arquivo = fopen(nomeArquivo, "w");
    if (!arquivo) {
        perror("Erro ao abrir arquivo para escrita");
        return;
    }

    No *atual = stack->topo;
    while (atual != NULL) {
        fprintf(arquivo, "%d\n", atual->valor.valor);
        atual = atual->prox;
    }

    fclose(arquivo);
}

File: Java/src/DynamicStack.java

public class DynamicStack {
    protected Node topo;
    protected int size;

    public DynamicStack() {
        this.topo = null;
        this.size = 0;
    }

    public boolean empilhar(Registro valor) {
        Node novo = new Node(valor);
        novo.prox = this.topo;
        this.topo = novo;
        this.size++;
        return true;
    }

    public Registro desempilhar() {
        if (pilhaVazia()) return null;

        Node remover = this.topo;
        Registro valor = remover.valor;
        this.topo = remover.prox;
        // Em Java, o garbage collector lida com a desalocação de memória
        // free(remover);
        this.size--;
        return valor;
    }

    public boolean pilhaVazia() {
        return this.size == 0;
    }

    public void imprimir() {
        Node atual = this.topo;
        while (atual != null) {
            System.out.print(atual.valor.valor + " ");
            atual = atual.prox;
        }
        System.out.println();
    }

    public void salvarEmArquivo(String nomeArquivo) {
        try (java.io.FileWriter writer = new java.io.FileWriter(nomeArquivo)) {
            Node atual = this.topo;
            while (atual != null) {
                writer.write(atual.valor.valor + "\n");
                atual = atual.prox;
            }
        } catch (java.io.IOException e) {
            System.err.println("Erro ao abrir arquivo para escrita: " + e.getMessage());
        }
    }
}

